pilot/pkg/config/kube/gateway/inferencepool.go

```diff
@@ -0,0 +1,408 @@
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	inferencev1alpha2 "sigs.k8s.io/gateway-api-inference-extension/api/v1alpha2"
+	gatewayv1 "sigs.k8s.io/gateway-api/apis/v1"
+	gateway "sigs.k8s.io/gateway-api/apis/v1beta1"
+
+	"istio.io/istio/pkg/config/constants"
+	"istio.io/istio/pkg/config/schema/gvk"
+	"istio.io/istio/pkg/config/schema/gvr"
+	"istio.io/istio/pkg/kube"
+	"istio.io/istio/pkg/kube/controllers"
+	"istio.io/istio/pkg/kube/kclient"
+)
+
+const maxServiceNameLength = 63
+const hashSize = 8
+const InferencePoolRefLabel = "inference.x-k8s.io/inference-pool-name"
+const InferencePoolExtensionRefSvc = "inference.x-k8s.io/extension-service"
+const InferencePoolExtensionRefPort = "inference.x-k8s.io/extension-port"
+
+// // ManagedLabel is the label used to identify resources managed by this controller
+// const ManagedLabel = "inference.x-k8s.io/managed-by"
+
+// ControllerName is the name of this controller for labeling resources it manages
+const ControllerName = "inference-controller"
+
+var supportedControllers = getSupportedControllers()
+
+func getSupportedControllers() sets.Set[gatewayv1.GatewayController] {
+	ret := sets.New[gatewayv1.GatewayController]()
+	for _, controller := range builtinClasses {
+		ret.Insert(controller)
+	}
+	return ret
+}
+
+// InferencePoolController implements a controller that materializes an InferencePool
+// into a Kubernetes Service to allow traffic to the inference pool.
+type InferencePoolController struct {
+	client     kube.Client
+	queue      controllers.Queue
+	pools      kclient.Client[*inferencev1alpha2.InferencePool]
+	services   kclient.Client[*corev1.Service]
+	httpRoutes kclient.Client[*gateway.HTTPRoute]
+	gateways   kclient.Client[*gateway.Gateway]
+	// ServiceToEndpointPickerService map[types.NamespacedName]*corev1.Service
+	clients map[schema.GroupVersionResource]getter
+}
+
+// NewInferencePoolController constructs a new InferencePoolController and registers required informers.
+func NewInferencePoolController(client kube.Client) *InferencePoolController {
+	filter := kclient.Filter{ObjectFilter: client.ObjectFilter()}
+	pools := kclient.NewFiltered[*inferencev1alpha2.InferencePool](client, filter)
+
+	ic := &InferencePoolController{
+		client:  client,
+		clients: map[schema.GroupVersionResource]getter{},
+		pools:   pools,
+	}
+
+	ic.queue = controllers.NewQueue("inference pool",
+		controllers.WithReconciler(ic.Reconcile),
+		controllers.WithMaxAttempts(5))
+
+	// Set up a handler that will add the parent InferencePool object onto the queue
+	parentHandler := controllers.ObjectHandler(controllers.EnqueueForParentHandler(ic.queue, gvk.InferencePool))
+
+	ic.services = kclient.NewFiltered[*corev1.Service](client, filter)
+	ic.services.AddEventHandler(parentHandler)
+	ic.clients[gvr.Service] = NewUntypedWrapper(ic.services)
+
+	ic.httpRoutes = kclient.NewFiltered[*gateway.HTTPRoute](client, filter)
```

Comment:
@ilrudie @howardjohn @danehans For some reason I dont get an event to enter the reconcile loop again when I change an HTTPRoute, although I register the parent handler. Have an idea why?

Reply:
oh, its because I dont really have an ownerRef nor I should have an ownerRef there. Would it be smart to store in memory all the inferencepool -> xRoutes and inferencePool -> gateways references? so I know what to queue?